Project description
Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children have an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.
Your report will be considered when building a credit score for a potential customer. A credit score is used to evaluate the ability of a potential borrower to repay their loan.
Instructions for completing the project
Step 1. Open the data file /datasets/credit_scoring_eng.csv and have a look at the general information.
Step 2. Preprocess the data:
Identify and fill in missing values
Replace the real number data type with the integer type
Delete duplicate data
Categorize the data
Be sure to explain:
Which missing values you identified
Possible reasons these missing values were present
Which method you used to fill in missing values
Which method you used to find and delete duplicate data and why
Possible reasons why duplicate data was present
Which method you used to change the data type and why
Which dictionaries you've selected for this dataset and why
The data may contain artifacts, or values that don't correspond to reality (for instance, a negative number of days employed). This kind of thing happens when you're working with real data. You need to describe the possible reasons such data may have turned up and process it.
Step 3. Answer these questions:
Is there a connection between having kids and repaying a loan on time?
Is there a connection between marital status and repaying a loan on time?
Is there a connection between income level and repaying a loan on time?
How do different loan purposes affect on-time loan repayment?
Interpret your answers. Explain what the results you obtained mean.
Step 4. Write an overall conclusion.
Format:
Complete the project in the Jupyter Notebook (it will open when you click Next). Write code in code cells and notes with explanations and interpretations in markdown cells. Use formatting and headings.
Description of the data
children: the number of children in the family
days_employed: how long the customer has been working
dob_years: the customer’s age
education: the customer’s education level
education_id: identifier for the customer’s education
family_status: the customer’s marital status
family_status_id: identifier for the customer’s marital status
gender: the customer’s gender
income_type: the customer’s income type
debt: whether the customer has ever defaulted on a loan
total_income: monthly income
purpose: reason for taking out a loan
How will my project be assessed?
Your reviewer will use assessment criteria to evaluate your project. Before getting started, read the criteria carefully. This will help ensure that you do a good job.
Here’s what mentors look at when assessing your project:
How do you describe the problems identified in the data?
What methods do you use to replace data types and to process missing values and duplicate data?
Did you categorize the data? Why did you do it the way you did?
Did you export the final data into pivot tables?
Did you use try-except clauses in your code to handle any unexpected errors?
Did you follow the correct project structure and write clean code?
Did you draw accurate and useful conclusions?
Did you leave comments on each step you took?
When you click "Let's go!" you will be taken to Practicum's JupyterHub. You can complete your project there.
Another option: you can install Python on your computer and do your project locally. There are many guides to installing Python and its libraries on your computer.
An easy way to install everything you need at once is to download the Anaconda distribution kit (https://www.anaconda.com/distribution/). During installation, check the following options in advanced settings:
Add Anaconda to the system PATH environment variable
Register Anaconda as the system Python 3.n (3.n will be replaced by the version number for the Python software included with your Anaconda download)
There are two ways to open the Jupyter Notebook:
Find the Jupyter Notebook in Anaconda Navigator and launch it by clicking the icon
Write jupyter notebook in Command Prompt (Windows) or Terminal (macOS)


try:
    df = pd.read_csv('datasets/real_estate_data_us.csv')
except:
    df = pd.read_csv('/datasets/real_estate_data_us.csv')
    


# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

In [1]:
# import libraries
import pandas as pd
import numpy as np

In [2]:
# load the data 
try:
    data = pd.read_csv('/Users/rraven/Desktop/a_final_yandex/datasets/credit_scoring_eng.csv')
except:
    data = pd.read_csv('/datasets/credit_scoring_eng.csv')

In [3]:
# print general info & first 10 rows
#data = pd.read_csv('../data/credit_scoring_eng.csv')
print('\nGeneral Info for credit_scoring_eng\n')
data.info()
print('\nFirst 10 rows of credit_scoring_eng.csv')
data.head(10)


General Info for credit_scoring_eng

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB

First 10 rows of credit_scoring_eng.csv


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [4]:
# find the number of missing values per column
print('Total Rows:', len(data))
print('\nColumn\t\tMissing Rows')
data.isnull().sum()

Total Rows: 21525

Column		Missing Rows


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
ratio = 100*data.isnull().sum() / len(data)
print('Column\t\tPercent Missing')
ratio

Column		Percent Missing


children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

In [6]:
# check data for duplicates in entire dataFrame
print('Number of duplicate rows:')
data.duplicated().sum()

Number of duplicate rows:


54

**Initial Observations for credit_scoring_eng.csv**

* 21525 rows
* 12 columns
* A mix of data types (float64(2), int64(5), object(5))
* 2174 missing values in 2 columns (days_employed, total_income) 
* days_employed and total_income are missing over 10% of total values
* 54 duplicate rows

In [7]:
# investigate children column
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [8]:
# investigate family_status column
data['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [9]:
# investigate family_status_id column
data['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

In [10]:
# investigate purpose column
data['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
buying property for renting out             653
housing transactions                        653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

In [11]:
# how many unique entries in purpose column
data['purpose'].nunique()

38

In [12]:
# investigate debt column
data['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

In [13]:
# investigate dob_years column
data['dob_years'].value_counts().sort_index()

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [14]:
# categorize dob_years in 10 year increments
less_than_20 = 0
btwn20and30 = 0
btwn30and40 = 0
btwn40and50 = 0
btwn50and60 = 0
btwn60and70 = 0
over_equal70 = 0
nan_count = 0
unknown = 0
for x in data['dob_years']:
    if x < 20:
        less_than_20 +=1
    elif x < 30:
        btwn20and30 +=1
    elif x < 40:
        btwn30and40 +=1   
    elif x < 50:
        btwn40and50 +=1 
    elif x < 60:
        btwn50and60 +=1 
    elif x < 70:
        btwn60and70 +=1 
    elif x >= 70:
        over_equal70 +=1        
    elif str(x) == 'nan':
        nan_count +=1
    else:
        # to check there are no other values
        unknown +=1
print("Age < 20:\t", less_than_20, 
      "\n20 <= Age < 30:\t", btwn20and30,
      "\n30 <= Age < 40:\t", btwn30and40,
      "\n40 <= Age < 50:\t", btwn40and50,
      "\n50 <= Age < 60:\t", btwn50and60,
      "\n60 <= Age < 70:\t", btwn60and70,
      "\nAge >= 70:\t", over_equal70,
      "\nnan:\t\t", nan_count,
      "\nUnknown:\t", unknown)

Age < 20:	 115 
20 <= Age < 30:	 3169 
30 <= Age < 40:	 5674 
40 <= Age < 50:	 5371 
50 <= Age < 60:	 4679 
60 <= Age < 70:	 2346 
Age >= 70:	 171 
nan:		 0 
Unknown:	 0


In [15]:
# investigate education column
data['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [16]:
# investigate education_id column
data['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

In [17]:
# investigate gender column
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [18]:
# investigate income_type column
data['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [19]:
# how many unique entries in income_type column
data['income_type'].nunique()

8

In [20]:
# investigate total_income
data['total_income'].value_counts()

17312.717    2
42413.096    2
31791.384    2
21005.772    1
18591.443    1
            ..
27020.895    1
23686.835    1
9606.294     1
28156.762    1
13127.587    1
Name: total_income, Length: 19348, dtype: int64

In [21]:
# categorize income_level into 10K levels to see distribution

# create a new function, income_level_fx
def income_level_fx(row):
    # the income_level is returned according to total_income
    income = row['total_income']
    
    if income < 10000:
        return 'less_than_10K' 
    elif income < 20000:
        return 'btwn10Kand20K'
    elif income < 30000:
        return 'btwn20Kand30K'   
    elif income < 40000:
        return 'btwn30Kand40K'  
    elif income < 50:
        return 'btwn40Kand50K' 
    elif income < 60000:
        return 'btwn50Kand60K'  
    elif income < 70000:
        return 'btwn60Kand70K' 
    elif income < 80000:
        return 'btwn70Kand80K' 
    elif income < 90000:
        return 'btwn80Kand90K' 
    elif income < 100000:
        return 'btwn90Kand100K'
    elif income >= 100000:
        return 'greater_than_100K'  
    
# create a new column, income_level, based on total_income
data['income_level'] = data.apply(income_level_fx, axis=1)

# verfiy new column: income_level
data['income_level'].value_counts()

btwn10Kand20K        6443
btwn20Kand30K        6063
btwn30Kand40K        3107
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

### Conclusion

**The datafile credit_scoring_eng.csv contains**
* 21525 rows
* 12 columns
* A mix of data types (float64(2), int64(5), object(5))
* 2174 missing values in 2 columns (days_employed, total_income) 
* days_employed and total_income are missing over 10% of total values
* 54 duplicate rows

**Targeted areas - having kids, marital status, income level and loan purpose r/t repayment**
* children 
    - int64
    - contains 47 entries with -1 and 76 with 20
    - maybe -1 was meant to be 1? keying error? correct or remove?
    - maybe 20 was meant to be 2? keying error? correct or remove?
    - ultimately can try different groups (none or some), (none, 1, or > 1), etc
* family_status and family_status_id 
    - string(object) and int64
    - no problems noted, appear to correlate well
        - married              12380 
        - civil partnership     4177
        - unmarried             2813
        - divorced              1195
        - widow / widower        960
        - 0    12380
        - 1     4177
        - 4     2813
        - 3     1195
        - 2      960
* total_income
    - float64
    - will need to address missing values    
* purpose
    - string(object)
    - 38 unique entries
    - appears to be 4 main categories 
        - wedding
        - property
        - car
        - education
    - will need to work with stem and/or lemmas
    
**How to judge risk of default / likelihood of repayment of loan?**
Discussion: There are 12 total columns. 5 of these are specifically involved in the task (children, family_status/family_status_id, total_income, and purpose). This leaves days_employed, dob_years, education,
education_id, gender, income_type, and debt.

Which of these are likely to correlate with risk of default? 
- days_employed might, because it could demonstrate stability, but the values in that column have too many problems to be useful. Therefore, that column will likely be ignored. 
- dob_years: could younger people (< 30 or 40) have a higher risk of default? maybe
- education / education_id: could less education correlate with higher risk of default? maybe
- gender: could gender correlate with risk of default? maybe
- income_type: could source of income correlate with risk of default? maybe
- debt: does "whether the customer has ever defaulted on a loan" correlate with risk of default? very likely

**Of these choices, debt is the most likely to produce useful correlations with risk of default. **
  
* debt
    - int64
    - 2 values, no nan or unknowns
    - assume 0 hasn't ever defaulted 
    - assume 1 for has defaulted in the past
        - 0    19784
        - 1     1741
     
**General observations of other columns**
* days_employed (will investigate in missing info section)
    - float64 (why?) 
    - will need to address missing values     
* dob_years
    - int64
    - age range from 19 to 75, but 101 values of 0 for age
    - will need to categorize, maybe by decade 
* education has a mixture of upper and lower case strings, numerous variations
    - string(object)
    - change all to lowercase to unify and compare to education_id
    - could use it to help fill in missing income info
* eduction_id column
    - int64
    - could be used instead of education if codes correlate?
        - 0 = bachelor's degree
        - 1 = secondary education
        - 2 = some college
        - 3 = primary education
        - 4 = graduate degree
    - maybe could reorganize because this order doesn't make sense?
* gender
    - string(object)
    - could change it to number code
* income_type
    - string(object)
    - 8 unique entries
    - could change it to number code
    - could combine it with other info to fill in missing incomes
        - employee                       11119
        - business                        5085
        - retiree                         3856
        - civil servant                   1459


## Data preprocessing

### Processing missing values

1. Address days_employed
2. Address total_income

In [22]:
# investigate days_employed column
data['days_employed'].value_counts().sort_index()

-18388.949901     1
-17615.563266     1
-16593.472817     1
-16264.699501     1
-16119.687737     1
                 ..
 401663.850046    1
 401674.466633    1
 401675.093434    1
 401715.811749    1
 401755.400475    1
Name: days_employed, Length: 19351, dtype: int64

In [23]:
# there are missing values, but how many?
# find percent of total rows where days_employed is missing a value
missing_days = data['days_employed'].isnull().sum()
pct_missing = missing_days/len(data)
print('Percentage of rows missing total_income: {:.2%}'.format(pct_missing))

Percentage of rows missing total_income: 10.10%


In [24]:
# add column & calculate years worked using days_employed
data['years_employed'] = data['days_employed'] / 365

In [25]:
# investigate range of years_employed column
data['years_employed'].value_counts().sort_index()

-50.380685      1
-48.261817      1
-45.461569      1
-44.560821      1
-44.163528      1
               ..
 1100.448904    1
 1100.477991    1
 1100.479708    1
 1100.591265    1
 1100.699727    1
Name: years_employed, Length: 19351, dtype: int64

There are some unrealistic values (working for negative 50 years and over 1100 years).

In [26]:
# find counts of positive and negative values for days employed
pos_count = 0
neg_count = 0
nan_count = 0
unknown = 0
for x in data['days_employed']:
    if x >= 0:
        pos_count +=1
    elif x < 0:
        neg_count +=1
    elif str(x) == 'nan':
        nan_count +=1
    else:
        # to check there are no other values
        unknown +=1
print("Positive days employed:\t", pos_count, 
      "\nNegative days employed:\t", neg_count,
      "\nnan in days employed:\t", nan_count,
      "\nUnknown in days employed:", unknown)

Positive days employed:	 3445 
Negative days employed:	 15906 
nan in days employed:	 2174 
Unknown in days employed: 0


Analysis of days_employed reveals:
- float64 (why?) and has many negative values (why?)
- is missing 2174 values
- has a range from -18388.949901 to 401755.400475 days employed
- has a range from -50 years to over 1100 years employed
- positive days employed: 3445 
- negative days employed: 15906 

Possible reasons for problem entries:
1. Human error when entering data (adding - by accident).
2. Error in units (perhaps the person entering the data got confused between hours worked and days worked?). Confusing hours for days may explain the very high numbers.
3. There may have been errors when merging different data sets.

Plan to fill missing values:
1. Since negative values are likely due to human error when entering data, use absolute value on days_employed.
2. Evaluate min/max values to determine range again, verify no more negative values.
3. Refresh years_employed and verify no more negative values.
4. Compare mean and median to gain an overall sense of data.
5. If people mistakenly entered hours worked instead of days, we can assume the maximum work period could be 70 years (70 yrs * 365 days/yrs) = 25550 days.
6. More realistically, we could assume the maximum work period could be 50 years (50 yrs * 365 days/yrs) = 18250.
7. With values greater than work period (either 25550 days or 18250 days), divide by 24 to get a days worked: data_50_yrs and data_70_years.
8. Compare statistics data_50_yrs and data_70_years.
9. Replace values in data df with appropriate cutoff value (#/24).
10. Compare mean and median of days_employed column.
11. Fill missing values with appropriate value (mean or median).

In [27]:
# apply abs() to change negative values to positive
data['days_employed'] = data['days_employed'].abs()
# verify no more negative values after abs()
negative_count = data.loc[data['days_employed']  < 0, 'days_employed'].count()
print('After applying abs(), verify the total number of negative values in days_employed = 0.')
negative_count

After applying abs(), verify the total number of negative values in days_employed = 0.


0

In [28]:
positive_count = data.loc[data['days_employed']  > 0, 'days_employed'].count()
print('After applying abs(), verify the total number of positive values in days_employed = 19351')
positive_count

After applying abs(), verify the total number of positive values in days_employed = 19351


19351

In [29]:
# add column & calculate years worked using days_employed
data['years_employed'] = data['days_employed'] / 365

days_neg = data.loc[data['years_employed']  < 0, 'years_employed'].count()
print('After applying abs(), verify the total number of negative values in years_employed = 0.')
days_neg

After applying abs(), verify the total number of negative values in years_employed = 0.


0

In [30]:
# find info on mean, min, max for days_employed
print('Statistical info for days_employed')
data['days_employed'].describe()

Statistical info for days_employed


count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [31]:
# find info on mean, min, max for years_employed
print('Statistical info for years_employed')
data['years_employed'].describe()

Statistical info for years_employed


count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: years_employed, dtype: float64

In [32]:
# find the median of years_employed
data['years_employed'].median()

6.0115631969279315

This data is skewed with some very high values. The median for years is just over 6, while the mean is over 183 years worked. This suggests there are many very high values distorting the data.

In [33]:
# make a copy of data, name it data_50_yrs
data_50_yrs = data.copy()
# change all the values greater than 18250 days (50 years) by dividing by 24 and saving results
data_50_yrs.loc[data_50_yrs['days_employed'] > 18250, 'days_employed'] = data_50_yrs['days_employed']/24
# find info on mean, min, max for days employed where values changed
print('Statistical info for days_employed')
print('where values > 18250 (50 years) and divided by 24')
data_50_yrs['days_employed'].describe()

Statistical info for days_employed
where values > 18250 (50 years) and divided by 24


count    19351.000000
mean      4640.730487
std       5355.124832
min         24.141633
25%        926.925347
50%       2194.216968
75%       5535.883055
max      17615.563266
Name: days_employed, dtype: float64

In [34]:
# make a copy of data, name it data_70_yrs
data_70_yrs = data.copy()
# change all the values greater than 25550 days (70 years) by dividing by 24 and saving results
data_70_yrs.loc[data_70_yrs['days_employed'] > 25550, 'days_employed'] = data_70_yrs['days_employed']/24
# find info on mean, min, max for days employed where values changed
print('Statistical info for days_employed')
print('where values > 25550 (70 years) and divided by 24')
data_70_yrs['days_employed'].describe()

Statistical info for days_employed
where values > 25550 (70 years) and divided by 24


count    19351.000000
mean      4641.641176
std       5355.964289
min         24.141633
25%        927.009265
50%       2194.220567
75%       5537.882441
max      18388.949901
Name: days_employed, dtype: float64

The means and medians for the two options are very close:

- 50 yrs  70 yrs
- 4640 vs 4641
- 2194 vs 2194

The max values are a bit different
17615 (48 years) vs 18388 (50 years).

The 70 year cutoff will be used in the data df just in case that entry with a 50 year work history is accurate. There are 3800+ retirees in the sample after all.

In [35]:
# change all the values greater than 25550 days (70 years) by dividing by 24 and saving results
data.loc[data['days_employed'] > 25550, 'days_employed'] = data['days_employed']/24

# find info on mean, min, max for days_employed
print('Statistical info for days_employed')
data['days_employed'].describe()

Statistical info for days_employed


count    19351.000000
mean      4641.641176
std       5355.964289
min         24.141633
25%        927.009265
50%       2194.220567
75%       5537.882441
max      18388.949901
Name: days_employed, dtype: float64

In [36]:
# look for nan, negative, and other values in days_employed
# print first 20 rows of duplicates to look for patterns
pos_count = 0
neg_count = 0
nan_count = 0
unknown = 0
counter= -1
for x in data['days_employed']:
    counter +=1
    if x >= 0:
        pos_count +=1
    elif x < 0:
        neg_count +=1
    elif str(x) == 'nan':
        nan_count +=1
        if nan_count < 20:
            print(data.iloc[counter])
    else:
        # to check there are no other values
        unknown +=1
print("Positive income:", pos_count, 
      "\nNegative income:", neg_count,
      "\nnan:\t\t", nan_count,
      "\nUnknown:\t", unknown)

children                              0
days_employed                       NaN
dob_years                            65
education           secondary education
education_id                          1
family_status         civil partnership
family_status_id                      1
gender                                M
income_type                     retiree
debt                                  0
total_income                        NaN
purpose               to have a wedding
income_level                       None
years_employed                      NaN
Name: 12, dtype: object
children                              0
days_employed                       NaN
dob_years                            41
education           secondary education
education_id                          1
family_status                   married
family_status_id                      0
gender                                M
income_type               civil servant
debt                                  0
total_income    

Name: 96, dtype: object
children                            0
days_employed                     NaN
dob_years                          47
education           bachelor's degree
education_id                        0
family_status                 married
family_status_id                    0
gender                              F
income_type                  employee
debt                                0
total_income                      NaN
purpose             profile education
income_level                     None
years_employed                    NaN
Name: 97, dtype: object
children                               0
days_employed                        NaN
dob_years                             46
education            secondary education
education_id                           1
family_status                    married
family_status_id                       0
gender                                 F
income_type                     employee
debt                                   0
total_inco

In [37]:
# find median of days_employed
median_days = data['days_employed'].median()
print('The median of days_employed')
median_days

The median of days_employed


2194.220566878695

In [38]:
# fill in missing values with median income
data['days_employed'].fillna(value=median_days, inplace = True)

In [39]:
# check for any missing values
# find info on mean, min, max for days_employed
print('Statistical info for days_employed')
data['days_employed'].describe()

Statistical info for days_employed


count    21525.000000
mean      4394.454537
std       5131.551830
min         24.141633
25%       1025.608174
50%       2194.220567
75%       4779.587738
max      18388.949901
Name: days_employed, dtype: float64

There are no more missing values and the median remains the same, 2194. The mean actually went down, which makes sense because over 2000 missing values were replaced with 2194, bringing the overall mean downward.

In [40]:
# find percent of total rows where total_income is missing
missing_income = data['total_income'].isnull().sum()
pct_missing = missing_income/len(data)
print('Percentage of rows missing total_income: {:.2%}'.format(pct_missing))

Percentage of rows missing total_income: 10.10%


In [41]:
# look for nan, negative, and other values in total_income
# print first 20 rows of duplicates to look for patterns
pos_count = 0
neg_count = 0
nan_count = 0
unknown = 0
counter= -1
for x in data['total_income']:
    counter +=1
    if x >= 0:
        pos_count +=1
    elif x < 0:
        neg_count +=1
    elif str(x) == 'nan':
        nan_count +=1
        if nan_count < 20:
            print(data.iloc[counter])

    else:
        # to check there are no other values
        unknown +=1
print("Positive income:", pos_count, 
      "\nNegative income:", neg_count,
      "\nnan:\t\t", nan_count,
      "\nUnknown:\t", unknown)

children                              0
days_employed               2194.220567
dob_years                            65
education           secondary education
education_id                          1
family_status         civil partnership
family_status_id                      1
gender                                M
income_type                     retiree
debt                                  0
total_income                        NaN
purpose               to have a wedding
income_level                       None
years_employed                      NaN
Name: 12, dtype: object
children                              0
days_employed               2194.220567
dob_years                            41
education           secondary education
education_id                          1
family_status                   married
family_status_id                      0
gender                                M
income_type               civil servant
debt                                  0
total_income    

Name: 135, dtype: object
children                              0
days_employed               2194.220567
dob_years                            39
education           secondary education
education_id                          1
family_status         civil partnership
family_status_id                      1
gender                                M
income_type                    employee
debt                                  0
total_income                        NaN
purpose                wedding ceremony
income_level                       None
years_employed                      NaN
Name: 141, dtype: object
children                              0
days_employed               2194.220567
dob_years                            62
education           secondary education
education_id                          1
family_status                   married
family_status_id                      0
gender                                M
income_type                     retiree
debt                          

In [42]:
# categorize income_level into 10K levels

# create a new function, income_level_fx
def income_level_fx(row):
    # the income_level is returned according to total_income
    income = row['total_income']
    
    if income < 10000:
        return 'less_than_10K' 
    elif income < 20000:
        return 'btwn10Kand20K'
    elif income < 30000:
        return 'btwn20Kand30K'   
    elif income < 40000:
        return 'btwn30Kand40K'  
    elif income < 50:
        return 'btwn40Kand50K' 
    elif income < 60000:
        return 'btwn50Kand60K'  
    elif income < 70000:
        return 'btwn60Kand70K' 
    elif income < 80000:
        return 'btwn70Kand80K' 
    elif income < 90000:
        return 'btwn80Kand90K' 
    elif income < 100000:
        return 'btwn90Kand100K'
    elif income >= 100000:
        return 'greater_than_100K'  
    
# create a new column, income_level, based on total_income
data['income_level'] = data.apply(income_level_fx, axis=1)

# verfiy new column: income_level
data['income_level'].value_counts()

btwn10Kand20K        6443
btwn20Kand30K        6063
btwn30Kand40K        3107
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

In [43]:
# find info on mean, min, max for total_income
print('Statistical info for total_income')
data['total_income'].describe()

Statistical info for total_income


count     19351.000000
mean      26787.568355
std       16475.450632
min        3306.762000
25%       16488.504500
50%       23202.870000
75%       32549.611000
max      362496.645000
Name: total_income, dtype: float64

In [44]:
# find median of total_income
median_income = data['total_income'].median()
print('The median of total_income')
median_income

The median of total_income


23202.87

total_income is vital to the analysis. Over 10% of the values are missing. 
Those values could be replaced by the mean or median. 

Since the mean (26787) > median (23202), high outling values are pulling the mean up. The median should be used to replace the values. A copy of the dataframe, titled data_median_income, will be created to store data where the median total_income replaces the missing values. 

Another option will be to look at correlations between other factors and total_income and replace missing values based on those factors. 

The two options will be compared in the final analysis.

In [45]:
# make a copy of data and replace the missing values with median
data_median_income = data.copy()

In [46]:
# fill in missing values with median income
data_median_income['total_income'].fillna(value=median_income, inplace = True)

In [47]:
# print the number of missing values in data_median_income after filling nan
print('Verify that there are no missing values in data_median_income')
data_median_income['total_income'].isnull().sum()

Verify that there are no missing values in data_median_income


0

In [48]:
# categorize data_median_income income_level into 10K levels

# create a new function, income_level_fx
def income_level_fx(row):
    # the income_level is returned according to total_income
    income = row['total_income']
    
    if income < 10000:
        return 'less_than_10K' 
    elif income < 20000:
        return 'btwn10Kand20K'
    elif income < 30000:
        return 'btwn20Kand30K'   
    elif income < 40000:
        return 'btwn30Kand40K'  
    elif income < 50:
        return 'btwn40Kand50K' 
    elif income < 60000:
        return 'btwn50Kand60K'  
    elif income < 70000:
        return 'btwn60Kand70K' 
    elif income < 80000:
        return 'btwn70Kand80K' 
    elif income < 90000:
        return 'btwn80Kand90K' 
    elif income < 100000:
        return 'btwn90Kand100K'
    elif income >= 100000:
        return 'greater_than_100K'  
    
# create a new column, income_level, based on total_income
data_median_income['income_level'] = data_median_income.apply(income_level_fx, axis=1)

# verfiy new column: income_level
data_median_income['income_level'].value_counts()

btwn20Kand30K        8237
btwn10Kand20K        6443
btwn30Kand40K        3107
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

In [49]:
# find info on mean, min, max for total_income
print('Statistical info for data_median_income')
data_median_income['total_income'].describe()

Statistical info for data_median_income


count     21525.000000
mean      26425.517984
std       15658.570891
min        3306.762000
25%       17247.708000
50%       23202.870000
75%       31286.979000
max      362496.645000
Name: total_income, dtype: float64

Since debt, children, family_status, purpose will be examined later, those will not be used when considering filling missing values in total_income.

Before considering relationships between total_income and (gender and/or education_id and/or dob_years and/or income_type), dob_years and education need to be addressed.

1. Process dob_years into categories (called age_group) 
2. Change the items in education to lowercase so it can be used for analysis.
3. Investigate gender, education, dob_years, income_type with total_income

In [50]:
# process dob_years into age_group

# create a new function, age_group_fx 
def age_group_fx(row):
    # the age group is returned according to dob_years
    age = row['dob_years']

    if age < 20:
        return 'less_than_20' 
    elif age < 30:
        return 'btwn20and30'
    elif age < 40:
        return 'btwn30and40'   
    elif age < 50:
        return 'btwn40and50' 
    elif age < 60:
        return 'btwn50and60' 
    elif age < 70:
        return 'btwn60and70' 
    elif age >= 70:
        return 'over_equal70'  
    
# create a new column, age_group, based on dob_years
data['age_group'] = data.apply(age_group_fx, axis=1)

# calculate number of items per age level
print('Number of items per age group')
data['age_group'].value_counts()

Number of items per age group


btwn30and40     5674
btwn40and50     5371
btwn50and60     4679
btwn20and30     3169
btwn60and70     2346
over_equal70     171
less_than_20     115
Name: age_group, dtype: int64

In [51]:
# calculate number of items per education column
data['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [52]:
# change items to lowercase
data['education'] = data['education'].str.lower()

# calculate number of items per education column
print('Number of items per education category')
data['education'].value_counts()

Number of items per education category


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [53]:
# compare with education #s with education_id #s
print('Number of items per education_id category')
data['education_id'].value_counts()

Number of items per education_id category


1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

The education column and the education_id column values match.

In [54]:
# investigate gender and total_income
data.groupby('gender').agg({'total_income': ['count', 'mean', 'median']})

total_income                         
              count          mean     median
gender                                      
F             12752  24655.604757  21464.845
M              6598  30907.144369  26834.295
XNA               1  32624.825000  32624.825

There is close to 5K difference, quite significant, between values for F or M.

Gender will be used to help fill in missing values for total income.

In [55]:
# investigate education and total_income
data.groupby('education').agg({'total_income': ['count', 'mean', 'median']})

total_income                          
                           count          mean      median
education                                                 
bachelor's degree           4716  33142.802434  28054.5310
graduate degree                6  27960.024667  25161.5835
primary education            261  21144.882211  18741.9760
secondary education        13693  24594.503037  21836.5830
some college                 675  29045.443644  25618.4640

There is a significant difference between incomes for different education levels.

Education will be used to help fill in missing values for total income.

In [56]:
# investigate count of missing total_income by education
print('Count of missing total income by education')
print(data[data['total_income'].isnull()]['education'].value_counts())

Count of missing total income by education
secondary education    1540
bachelor's degree       544
some college             69
primary education        21
Name: education, dtype: int64


In [57]:
# investigate age_group and total_income
data.groupby('age_group').agg({'total_income': ['count', 'mean', 'median']})

total_income                          
                    count          mean      median
age_group                                          
btwn20and30          2871  25572.630177  22799.2580
btwn30and40          5109  28312.479963  24667.5280
btwn40and50          4834  28551.375635  24764.2290
btwn50and60          4178  25811.700327  22203.0745
btwn60and70          2095  23242.812818  19817.4400
less_than_20          104  24291.556587  21100.0045
over_equal70          160  20125.658331  18751.3240

There is not as significant difference between incomes for different age groups. 

Age_groups will not be used to help fill in missing values for total income.

In [58]:
# investigate income_type and total_income
data.groupby('income_type').agg({'total_income': ['count', 'mean', 'median']})

total_income                          
                                   count          mean      median
income_type                                                       
business                            4577  32386.793835  27577.2720
civil servant                       1312  27343.729582  24071.6695
employee                           10014  25820.841683  22815.1035
entrepreneur                           1  79866.103000  79866.1030
paternity / maternity leave            1   8612.661000   8612.6610
retiree                             3443  21940.394503  18962.3180
student                                1  15712.260000  15712.2600
unemployed                             2  21014.360500  21014.3605

There is a sizeable difference between business, employee, and retiree and all of those have over 3K applicants.

Income_type will be used to help fill in missing values for total income.

In [59]:
# investigate count of missing total_income by income_type
print('Count of missing total income by income_type')
print(data[data['total_income'].isnull()]['income_type'].value_counts())

Count of missing total income by income_type
employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64


In [60]:
# investigate if there are any correlations with total_income and 
# gender and education and income_type and age group
data.pivot_table(values=['total_income'], columns=['gender', 'education', 'income_type'])

gender                       F                                           \
education    bachelor's degree                                            
income_type           business civil servant      employee entrepreneur   
total_income       34893.06224  28466.106994  28491.963702    79866.103   

gender                                                            \
education                             graduate degree              
income_type        retiree unemployed   civil servant    retiree   
total_income  26175.804276  32435.602       17822.757  40868.031   

gender                                        ...                   M  \
education    primary education                ... secondary education   
income_type           business civil servant  ...            business   
total_income      24474.135692     14339.034  ...        33021.827315   

gender                                                             \
education                                                           
income_type  civil servant      employee       retiree unemployed   
total_income   30151.25731  27486.725383  22526.824219   9593.119   

gender                                                                     XNA  
education    some college                                         some college  
income_type      business civil servant      employee     retiree     business  
total_income  33777.41098   31869.49925  33293.136649  25214.5752    32624.825  

[1 rows x 42 columns]

This suggests some wide variations in value when categorized by gender, education, and income type.

In [61]:
# fill in values using education_id and income_type
print('Fill in missing values of total_income based on education, gender, and income_type')
data['total_income'] = data['total_income'].fillna(data.groupby(['education_id', 'income_type', 'gender'])['total_income'].transform('median'))

# print the number of missing values in data after filling nan
print('Verify that there are no missing values in data')
data['total_income'].isnull().sum()

Fill in missing values of total_income based on education, gender, and income_type
Verify that there are no missing values in data


1

In [62]:
miss_value = data[data['total_income'].isna()]
miss_value

children  days_employed  dob_years          education  education_id  \
5936         0    2194.220567         58  bachelor's degree             0   

     family_status  family_status_id gender   income_type  debt  total_income  \
5936       married                 0      M  entrepreneur     0           NaN   

                          purpose income_level  years_employed    age_group  
5936  buy residential real estate         None             NaN  btwn50and60

1 row still has a missing value for total_income.
The applicant is a male entrepreneur with a bachelor's degree.

In [63]:
# investigate if there are any correlations with total_income and 
# gender and income_type
data.pivot_table(values=['total_income'], columns=['gender', 'income_type'])

gender                   F                                           \
income_type       business civil servant      employee entrepreneur   
total_income  29130.340809  24625.803881  23527.169018    79866.103   

gender                                                            \
income_type  paternity / maternity leave      retiree unemployed   
total_income                    8612.661  21134.46767  32435.602   

gender                   M                                            \
income_type       business civil servant      employee       retiree   
total_income  36774.923641  33715.596379  28704.088398  23849.082213   

gender                                   XNA  
income_type    student unemployed   business  
total_income  15712.26   9593.119  32624.825

Since there is not a listing for a M who is an entrepreneur, the program couldn't fill in a value. The value for a F who is an entrepreneur is very high, so it may be better to base the missing value on a M having a bachelor's degree.

In [64]:
# investigate if there are any correlations with total_income and 
# gender and education_id
data.pivot_table(values=['total_income'], columns=['gender', 'education'])

gender                       F                                    \
education    bachelor's degree graduate degree primary education   
total_income      29896.324842       29345.394      18977.534635   

gender                                                         M  \
education    secondary education  some college bachelor's degree   
total_income        22387.501393  26126.363111      38399.014771   

gender                                                              \
education    graduate degree primary education secondary education   
total_income        27267.34      23703.273277        28035.492257   

gender                              XNA  
education     some college some college  
total_income  32934.841247    32624.825

In [65]:
# replace the missing value with value for a M with a bachelor's degree
print('Fill in missing value of total_income based on education and gender')
data['total_income'] = data['total_income'].fillna(data.groupby(['education_id', 'gender'])['total_income'].transform('median'))

# print the number of missing values in data after filling nan
print('Verify that there are no missing values in data')
data['total_income'].isnull().sum()

Fill in missing value of total_income based on education and gender
Verify that there are no missing values in data


0

In [66]:
# categorize income_level into 10K levels
# important to do after all the missing values added to update

# create a new function, income_level_fx
def income_level_fx(row):
    # the income_level is returned according to total_income
    income = row['total_income']
    
    if income < 10000:
        return 'less_than_10K' 
    elif income < 20000:
        return 'btwn10Kand20K'
    elif income < 30000:
        return 'btwn20Kand30K'   
    elif income < 40000:
        return 'btwn30Kand40K'  
    elif income < 50:
        return 'btwn40Kand50K' 
    elif income < 60000:
        return 'btwn50Kand60K'  
    elif income < 70000:
        return 'btwn60Kand70K' 
    elif income < 80000:
        return 'btwn70Kand80K' 
    elif income < 90000:
        return 'btwn80Kand90K' 
    elif income < 100000:
        return 'btwn90Kand100K'
    elif income >= 100000:
        return 'greater_than_100K'  
    
# create a new column, income_level, based on total_income
data['income_level'] = data.apply(income_level_fx, axis=1)

# verfiy new column: income_level
data['income_level'].value_counts()

btwn10Kand20K        7269
btwn20Kand30K        7101
btwn30Kand40K        3417
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

In [67]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
 12  income_level      21525 non-null  object 
 13  years_employed    19351 non-null  float64
 14  age_group         21525 non-null  object 
dtypes: float64(3), int64(5), object(7)
memory usage: 2.5+ MB


In [68]:
# find info on mean, min, max for total_income
print('Statistical info for total_income')
data['total_income'].describe()

Statistical info for total_income


count     21525.000000
mean      26457.155496
std       15723.056300
min        3306.762000
25%       17234.448000
50%       23181.085000
75%       31461.575000
max      362496.645000
Name: total_income, dtype: float64

### Conclusion

1. days_employed may or may not be useful for analysis. Negative values may have occured through human data entry error. Very large values likely occured because of confusion over units (hours versus days), but they might be due to problems merging datasets. To handle the negative values, the absolute value was applied and the new values stored in place. Large values (equivalent to working > 70 years) were divided by 24 (for 24 hours) and replaced. The first 20 rows with NAN values revealed no particular pattern, so the missing values are likely MAR. Then the missing values were filled using the median because there were still large values skewing the results (mean 4641 vs median 2194).


2. total_income is vital to analysis (in fact it is one of the key categories to report on). Over 10% of the values are missing. The first 20 duplicates were printed and there is no obvious pattern, therefore these missing values are MAR. Reasons for missing values could include human error, information not provided by applicant, or a mix up when datasets were merged. total_income was categorized into 10K increments and general statistics were displayed. The mean is greater than the median, therefore the median values will be used for replacing missing values.

3. Replacing missing values:
- approach one is to replace those missing values with the median. data_median_income
- approach two involves filling missing values based on a composite value drawn from the influence of gender, income_type, and education (except for one stray value where only gender and education were used).
- these approaches will be compared in the final analysis.

### Data type replacement

1. total_income should be changed from float 64 to an int for visual appeal / ease of understanding.
2. Data types could be changed to conserve memory. This is especially useful for very large files.

In [69]:
# investigate data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
 12  income_level      21525 non-null  object 
 13  years_employed    19351 non-null  float64
 14  age_group         21525 non-null  object 
dtypes: float64(3), int64(5), object(7)
memory usage: 2.5+ MB


In [70]:
# check memory usage
print('Memory useage before')
data.memory_usage()

Memory useage before


Index                  128
children            172200
days_employed       172200
dob_years           172200
education           172200
education_id        172200
family_status       172200
family_status_id    172200
gender              172200
income_type         172200
debt                172200
total_income        172200
purpose             172200
income_level        172200
years_employed      172200
age_group           172200
dtype: int64

In [71]:
# calculate total memory usage before
print('Memory useage before in MB')
memory_before = data.memory_usage().sum() / (1024**2) #converting to megabytes
memory_before

Memory useage before in MB


2.4634628295898438

In [72]:
# change data types using astype and apply w/numpy
data['children'] = data['children'].astype('int16')
# converting days_employed took 2 steps, 1st numpy to int, then astype
data['days_employed'] = data['days_employed'].apply(np.int)
data['days_employed'] = data['days_employed'].astype('int16')
data['dob_years'] = data['dob_years'].astype('int16')
data['education'] = data['education'].astype('category')
data['education_id'] = data['education_id'].astype('int16')
data['family_status'] = data['family_status'].astype('category')
data['family_status_id'] = data['family_status_id'].astype('int16')
data['gender'] = data['gender'].astype('category')
data['income_type'] = data['income_type'].astype('category')
data['debt'] = data['debt'].astype('int16')
# converting total_income took 2 steps, 1st numpy to int, then astype
data['total_income'] = data['total_income'].apply(np.int)
data['total_income'] = data['total_income'].astype('int16')
data['purpose'] = data['purpose'].astype('category')
data['years_employed'] = data['years_employed'].astype('float32')
data['age_group'] = data['age_group'].astype('category')

In [73]:
# check memory usage after
print('Memory useage after')
data.memory_usage()

Memory useage after


Index                  128
children             43050
days_employed        43050
dob_years            43050
education            21737
education_id         43050
family_status        21737
family_status_id     43050
gender               21657
income_type          21889
debt                 43050
total_income         43050
purpose              22901
income_level        172200
years_employed       86100
age_group            21881
dtype: int64

In [74]:
# calculate total memory usage after
print('Memory useage after in MB')
memory_after = data.memory_usage().sum() / (1024**2) #converting to 
memory_change = memory_before - memory_after
memory_after

Memory useage after in MB


0.6595420837402344

### Conclusion

In [75]:
print('Changing data types saved', memory_change, 'MB of memory and changing total_income and days_employed to int type allows for easier reading. Data types were mostly changed using astype, but when changing from float to int apply.np needed to be used. It is good to know that apply.np can only be used on columns with no missing values.') 

Changing data types saved 1.8039207458496094 MB of memory and changing total_income and days_employed to int type allows for easier reading. Data types were mostly changed using astype, but when changing from float to int apply.np needed to be used. It is good to know that apply.np can only be used on columns with no missing values.


### Processing duplicates

1. Manage duplicate rows
- Calculate duplicate rows.
- There is no reason to check for duplicates within columns, as values can repeat.

2. Manage purpose column with stemming or lemmatization


In [76]:
dup_rows = data.duplicated().sum()
pct_duplicated = dup_rows/len(data)
print('There are', dup_rows, 'duplicate rows in the file')
print('Percentage of duplicate rows = {:.2%}'.format(pct_duplicated))

There are 71 duplicate rows in the file
Percentage of duplicate rows = 0.33%


In [77]:
# remove the duplicate rows
data = data.drop_duplicates()

In [78]:
# verify that there are no more duplicate rows
print('Number of duplicates after dropping:')
data.duplicated().sum()

Number of duplicates after dropping:


0

In [79]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21454 non-null  int16   
 1   days_employed     21454 non-null  int16   
 2   dob_years         21454 non-null  int16   
 3   education         21454 non-null  category
 4   education_id      21454 non-null  int16   
 5   family_status     21454 non-null  category
 6   family_status_id  21454 non-null  int16   
 7   gender            21454 non-null  category
 8   income_type       21454 non-null  category
 9   debt              21454 non-null  int16   
 10  total_income      21454 non-null  int16   
 11  purpose           21454 non-null  category
 12  income_level      21454 non-null  object  
 13  years_employed    19351 non-null  float32 
 14  age_group         21454 non-null  category
dtypes: category(6), float32(1), int16(7), object(1)
memory usage: 840.6+ K

In [80]:
# use stemming to categorize purpose column
from nltk.stem import SnowballStemmer

english_stemmer = SnowballStemmer('english')   

data['purpose_words'] = data['purpose'].str.split().apply(lambda x: [english_stemmer.stem(y) for y in x])

def purpose_group(purpose_words):

    if 'wed' in purpose_words:
        return 'wedding'
    elif 'estat' in purpose_words or 'hous' in purpose_words or 'properti' in purpose_words:
        return 'real_estate'
    elif 'car' in purpose_words:
        return 'car'
    else:# 'educ' or 'uni' in purpose_words:
        return 'education'
    
data['purpose_cat'] = data['purpose_words'].apply(purpose_group)
data.head(10)

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0          14177         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0        -24916   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0        -22716   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0        -24614   
6            married                 0      F    business     0        -27052   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose   income_level  years_employed  \
0                purchase of the house  btwn50Kand60K       23.116913   
1                         car purchase  btwn10Kand20K       11.026859   
2                purchase of the house  btwn20Kand30K       15.406637   
3              supplementary education  btwn50Kand60K       11.300677   
4                    to have a wedding  btwn20Kand30K      932.235840   
5                purchase of the house  btwn50Kand60K        2.537495   
6                 housing transactions  btwn30Kand40K        7.888225   
7                            education  btwn20Kand30K        0.418574   
8                     having a wedding  btwn10Kand20K       18.985931   
9  purchase of the house for my family  btwn20Kand30K        5.996593   

     age_group                              purpose_words  purpose_cat  
0  btwn40and50                   [purchas, of, the, hous]  real_estate  
1  btwn30and40                             [car, purchas]          car  
2  btwn30and40                   [purchas, of, the, hous]  real_estate  
3  btwn30and40                      [supplementari, educ]    education  
4  btwn50and60                         [to, have, a, wed]      wedding  
5  btwn20and30                   [purchas, of, the, hous]  real_estate  
6  btwn40and50                           [hous, transact]  real_estate  
7  btwn50and60                                     [educ]    education  
8  btwn30and40                             [have, a, wed]      wedding  
9  btwn40and50  [purchas, of, the, hous, for, my, famili]  real_estate

In [81]:
print('Purpose        Count')
data['purpose_cat'].value_counts()

Purpose        Count


real_estate    10811
car             4306
education       4013
wedding         2324
Name: purpose_cat, dtype: int64

In [82]:
print('Total number of purpose values')
len(data['purpose_cat'])

Total number of purpose values


21454

### Conclusion

1. Duplicate rows increased from 54 to 71 after filling in missing values for total_income, but that is still only 0.33%. Duplicate rows can happen when datasets are merged or through human error.
2. Duplicate rows were deleted using drop_duplicates() since it is simple to use. Duplicate removal verified.
3. 4 categories (real_estate, car, education, wedding) used to filter purpose column.
4. Total number of values verified (each purpose_cat assigned a category)

### Categorizing Data

Categorizing data stratifies a large collection of values into groups or levels. It is ideal to use when working with age, income, time or anything that could be continuous but needs to be examined in groups.

1. Verify total_income properly categorized income_level for both data df and data_median_income df. See section 1.1 for initial categorizing total_income into income_level.
2. Categorize children into child_groups.
3. See section 1.1 for categorizing dob_years into age_group

In [83]:
# verfiy new column: income_level
data['income_level'].value_counts()

btwn10Kand20K        7220
btwn20Kand30K        7082
btwn30Kand40K        3414
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

In [84]:
# verfiy new column: income_level
data_median_income['income_level'].value_counts()

btwn20Kand30K        8237
btwn10Kand20K        6443
btwn30Kand40K        3107
btwn50Kand60K        2140
less_than_10K         926
btwn60Kand70K         294
btwn70Kand80K         156
greater_than_100K      99
btwn80Kand90K          83
btwn90Kand100K         40
Name: income_level, dtype: int64

In [85]:
# create new column based on no children, 1 child, 2 or more children
# first investigate problem values
data['children'].value_counts()

 0     14091
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [86]:
# calculate percentage of problem values (-1)
print('Percent of erroneous values (-1)')
print(47/21454)

Percent of erroneous values (-1)
0.0021907336627202388


In [87]:
# change the -1 values to 1 
data.loc[data['children'] < 0, ['children']] = 1

In [88]:
# calculate percentage of problem values (20)
print('Percent of likely erroneous values (20)')
print(76/21454)

Percent of likely erroneous values (20)
0.003542462943973152


In [89]:
# change the -1 values to 1 
data.loc[data['children'] == 20, ['children']] = 2

In [90]:
data['children'].value_counts()

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

Replaced likely erroneous values (1 for -1) and (2 for 20).

In [91]:
# create a new function, children_fx
def children_fx(row):
    # the income_level is returned according to total_income
    child = row['children']
    
    if child == 0:
        return 'no children'
    if child == 1:
        return '1 child'
    else:
        return '2 or more children'
    
# create a new column, income_level, based on total_income
data['child_groups'] = data.apply(children_fx, axis=1)
print('New categories for child groups')
data['child_groups'].value_counts()

New categories for child groups


no children           14091
1 child                4855
2 or more children     2508
Name: child_groups, dtype: int64

In [92]:
# verify child_groups added 
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0          14177         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0        -24916   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0        -22716   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose   income_level  years_employed    age_group  \
0    purchase of the house  btwn50Kand60K       23.116913  btwn40and50   
1             car purchase  btwn10Kand20K       11.026859  btwn30and40   
2    purchase of the house  btwn20Kand30K       15.406637  btwn30and40   
3  supplementary education  btwn50Kand60K       11.300677  btwn30and40   
4        to have a wedding  btwn20Kand30K      932.235840  btwn50and60   

              purpose_words  purpose_cat        child_groups  
0  [purchas, of, the, hous]  real_estate             1 child  
1            [car, purchas]          car             1 child  
2  [purchas, of, the, hous]  real_estate         no children  
3     [supplementari, educ]    education  2 or more children  
4        [to, have, a, wed]      wedding         no children

In [93]:
# verify no missing values
print('Total number of entries')
data['child_groups'].count()

Total number of entries


21454

### Conclusion

1. income_level verfied for both data df and data_median_income df. 
2. Replaced -1 and 20 values in children, grouped into 3 categories in child_groups. Verified no missing values.

At this point, clean up df and get rid of cols not used

In [94]:
# remove days_employed and years_employed columns
del data['days_employed'] 
del data['years_employed']
del data['dob_years']
del data['purpose_words']
del data['education_id']
del data['family_status_id']
data.head()

children            education      family_status gender income_type  debt  \
0         1    bachelor's degree            married      F    employee     0   
1         1  secondary education            married      F    employee     0   
2         0  secondary education            married      M    employee     0   
3         3  secondary education            married      M    employee     0   
4         0  secondary education  civil partnership      F     retiree     0   

   total_income                  purpose   income_level    age_group  \
0        -24916    purchase of the house  btwn50Kand60K  btwn40and50   
1         17932             car purchase  btwn10Kand20K  btwn30and40   
2         23341    purchase of the house  btwn20Kand30K  btwn30and40   
3        -22716  supplementary education  btwn50Kand60K  btwn30and40   
4         25378        to have a wedding  btwn20Kand30K  btwn50and60   

   purpose_cat        child_groups  
0  real_estate             1 child  
1          car             1 child  
2  real_estate         no children  
3    education  2 or more children  
4      wedding         no children

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [95]:
# create a formatting rule for ease of visualization
def format_float(value):
    value = value*100
    return f'{value:,.2f}%'
pd.options.display.float_format = format_float

In [96]:
# create pivot table with percents per child group category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['child_groups'])

Pecentage of applicants with a history of default


child_groups  1 child  2 or more children  no children
debt            9.17%               9.29%        7.54%

### Conclusion

Yes, there is a clear relationship between debt and no children versus debt and any children. Only 7.54% of applicants without a child defaulted, while appliants with 1 or more children defaulted at a rate of 9.17% to 9.29%. Therefore, a applicant with a child may pose a greater default risk.

- Is there a relation between marital status and repaying a loan on time?

In [97]:
# create pivot table with percents per marital status category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['family_status'])

Pecentage of applicants with a history of default


family_status  civil partnership  divorced  married  unmarried  \
debt                       9.35%     7.11%    7.55%      9.75%   

family_status  widow / widower  
debt                     6.57%

### Conclusion

Yes, there are differences in default history amoung applicants with different family status. Widowers pose the best risk, as only 6.57% of them defaulted on a loan in the past. Both married and divorced applicants pose a moderate risk (between 7.11% and 7.55% defaulted in the past). The biggest risk is for unmarried applicants (9.75%) and those in civil partnerships (9.35%). Therefore, appliants who are unmarried or in a civil partnership may pose the greatest default risk.

- Is there a relation between income level and repaying a loan on time?

In [98]:
# create pivot table with percents per income level category
print('Pecentage of applicants with a history of default')
print('for per income level where missing values replaced')
print('based on gender, education level, and income type.')
data.pivot_table(values=['debt'], columns=['income_level'])

Pecentage of applicants with a history of default
for per income level where missing values replaced
based on gender, education level, and income type.


income_level  btwn10Kand20K  btwn20Kand30K  btwn30Kand40K  btwn50Kand60K  \
debt                  8.50%          8.75%          7.47%          7.29%   

income_level  btwn60Kand70K  btwn70Kand80K  btwn80Kand90K  btwn90Kand100K  \
debt                  5.44%          5.13%          7.23%           5.00%   

income_level  greater_than_100K  less_than_10K  
debt                      6.06%          6.26%

In [99]:
# create pivot table with percents per income level category
print('Pecentage of applicants with a history of default')
print('for per income level (where missing values replaced')
print('by median value')
data_median_income.pivot_table(values=['debt'], columns=['income_level'])

Pecentage of applicants with a history of default
for per income level (where missing values replaced
by median value


income_level  btwn10Kand20K  btwn20Kand30K  btwn30Kand40K  btwn50Kand60K  \
debt                  8.54%          8.46%          7.79%          7.29%   

income_level  btwn60Kand70K  btwn70Kand80K  btwn80Kand90K  btwn90Kand100K  \
debt                  5.44%          5.13%          7.23%           5.00%   

income_level  greater_than_100K  less_than_10K  
debt                      6.06%          6.26%

### Conclusion

Interestingly, the percentages for both df (the one where 1 median value replaced all missing values and the one where gender, education level, and income type guided replacement values) are very similar.

Yes, there does seem to be a general trend where the higher the income, the less likely an applicant has defaulted in the past. Those making below60K demonstrate a clear trend towards higher risk as the income drops (7.29%, 7.79%, 8.46%, and then 8.54% for those making less than 20K. The upper half (greater than 60K) defaulted at a rate between 5.13% and 7.23%). Overall, those applicants earning less than 60K may need greater scrutiny as there is a higher likelyhood that they defaulted in the past.

- How do different loan purposes affect on-time repayment of the loan?

In [100]:
# create pivot table with percents per purpose category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['purpose_cat'])

Pecentage of applicants with a history of default


purpose_cat   car  education  real_estate  wedding
debt        9.36%      9.22%        7.23%    8.00%

### Conclusion

Yes, once again there is a relationship between the purpose of the loan and a history of default. Applicants wishing to purchase a car have the greatest historical defalut rate (9.36%), followed by those who want money for education (9.22%). Applicants looking to buy real estate pose the least risk, as only 7.23% defaulted in the past.


Investigate other columns for potential relationships to guide future analysis.

In [101]:
# create pivot table with percents per age group category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['age_group'])

Pecentage of applicants with a history of default


age_group  btwn20and30  btwn30and40  btwn40and50  btwn50and60  btwn60and70  \
debt            10.99%        9.75%        7.55%        6.55%        5.02%   

age_group  less_than_20  over_equal70  
debt              7.83%         3.55%

Age group: Looking into age may be worthwhile as there appears to be a sharp increase in historical default for those under 40.

In [102]:
# create pivot table with percents per gender category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['gender'])

Pecentage of applicants with a history of default


gender     F      M   XNA
debt   7.01% 10.26% 0.00%

Gender: It appears male applicants have a higher rate of default in this sample. Further analysis may be useful.

In [103]:
# create pivot table with percents per income_type category
print('Pecentage of applicants with a history of default')
data.pivot_table(values=['debt'], columns=['income_type'])

Pecentage of applicants with a history of default


income_type  business  civil servant  employee  entrepreneur  \
debt            7.40%          5.90%     9.57%         0.00%   

income_type  paternity / maternity leave  retiree  student  unemployed  
debt                             100.00%    5.64%    0.00%      50.00%

In [104]:
data['income_type'].value_counts()

employee                       11084
business                        5078
retiree                         3829
civil servant                   1457
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

Income type: Only the 4 groups (employee, business, retiree, civil servant) with the greatest number of applicants could be used for analysis since a sample of 1 or 2 isn't useful. Still, there does seem to be a significant difference between the lowest risk (retiree 5.64%) and highest risk (employee 9.57%). It would be worth more analysis.

## General conclusion

Number of children, family status, income, and the stated purpose of the loan can be used to increase the validity of the credit scoring system. 

Better loan risk (lower percentage of historical defaults on loans):
- no children
- widowers
- married
- divorced
- income > 60K
- purpose of loan: real estate
- purpose of loan: wedding

Higher loan risk (higher percentage of historical defaults on loans):
- have children
- unmarried
- in a civil union
- income < 60K
- purpose of loan: education
- purpose of loan: car 

Additionally, it may be advantageous to pursue analysis of age group, gender, and income type as there do appear to be relationships beween subsections and percentage of defaults.